File: 3070. count-submatrices-with-top-left-element-and-sum-less-than-k.py
class Solution:
    def countSubmatrices(self, grid: list[list[int]], k: int) -> int:
        m, n = len(grid), len(grid[0])
        ans = 0
        px = [[0] * (n + 1) for _ in range(m + 1)]
        for i in range(m):
                if px[i+1][j+1] <= k:
                    ans += 1
        return ans

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1:
    grid = [[7,6,3],[6,6,1]]
    k = 18
    output = sol.countSubmatrices(grid, k)
    print(f"Example 1: Input: grid = {grid}, k = {k}")
    print(f"Output: {output}")
    print(f"Expected: 4")
    print()
    
    # Example 2:
    grid = [[7,2,9],[1,5,0],[2,6,6]]
    k = 20
    output = sol.countSubmatrices(grid, k)
    print(f"Example 2: Input: grid = {grid}, k = {k}")
    print(f"Output: {output}")
    print(f"Expected: 6")

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1:
    grid = [[7,6,3],[6,6,1]]
    k = 18
    output = sol.countSubmatrices(grid, k)
    print(f"Example 1: Input: grid = {grid}, k = {k}")
    print(f"Output: {output}")
    print(f"Expected: 4")
    print()
    
    # Example 2:
    grid = [[7,2,9],[1,5,0],[2,6,6]]
    k = 20
    output = sol.countSubmatrices(grid, k)
    print(f"Example 2: Input: grid = {grid}, k = {k}")
    print(f"Output: {output}")
    print(f"Expected: 6")

File: 3070. count-submatrices-with-top-left-element-and-sum-less-than-k.java
class Solution {
    public int countSubmatrices(int[][] grid, int k) {
        int m = grid.length, n = grid[0].length;
        int ans = 0;
        int[][] px = new int[m + 1][n + 1];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                px[i+1][j+1] = grid[i][j] + px[i][j+1] + px[i+1][j] - px[i][j];
                if (px[i+1][j+1] <= k) ans++;
            }
        }
        return ans;
    }
}
	
	public static void main(String[] args) {
		Solution sol = new Solution();
		
		// Example 1:
		int[][] grid1 = {{1,2,3},{4,5,6},{7,8,9}};
		int k1 = 4;
		int output1 = sol.countSubmatrices(grid1, k1);
		System.out.println("Example 1: Input: grid = [[1,2,3],[4,5,6],[7,8,9]], k = " + k1);
		System.out.println("Output: " + output1);
		System.out.println("Expected: 4");
		System.out.println();
		
		// Example 2:
		int[][] grid2 = {{2}};
		int k2 = 3;
		int output2 = sol.countSubmatrices(grid2, k2);
		System.out.println("Example 2: Input: grid = [[2]], k = " + k2);
		System.out.println("Output: " + output2);
		System.out.println("Expected: 1");
	}
}